src/watchman_client.js

```diff
@@ -0,0 +1,341 @@
+    _client.on('error', onError.bind(self));
+    _client.on('end', onEnd.bind(self));
+  }
+
+  function getSubscription() {
+    let val = 'sane_' + _subscriptionId++;
+    return val;
+  }
+
+  /**
+   * Create a new watcherInfo entry for the given watchmanWatcher and 
+   * initialize it.
+   */
+  function createWatcherInfo(watchmanWatcher) {
+    let watcherInfo = {
+      subscription: getSubscription(),
+      watchmanWatcher: watchmanWatcher,           
+      root: null,                      // set during 'watch' or 'watch-project'
+      relativePath: null,              // same
+      since: null,                     // set during 'clock'
+      options: null                    // created and set during 'subscribe'.
+    };
+
+    _watcherMap[watcherInfo.subscription] = watcherInfo;
+
+    return watcherInfo;
+  }
+
+  /**
+   * Find an existing watcherInfo instance.
+   */
+  function getWatcherInfo(subscription) {
+    var watcherInfo = _watcherMap[subscription];
+    return watcherInfo;
+  }
+
+  /**
+   * Given a watchmanWatcher and a root, issue the correct 'watch'
+   * or 'watch-project' command and handle it with the callback.
+   * Because we're operating in 'sane', we'll keep the results
+   * of the 'watch' or 'watch-project' here.
+   */
+  function watch(subscription, root) {
+    var self = this;
+    
+    let promise = new Promise((resolve, reject) => {
+      let watcherInfo = getWatcherInfo(subscription);
+
+      if (_relative_root) { 
+        _client.command(['watch-project', root], (error, resp) => {
+          if (error) {
+            reject(error);
+          } else {
+            watcherInfo.root = resp.watch;
+            watcherInfo.relativePath = resp.relative_path ? resp.relative_path : '';
+            resolve(resp);
+          }
+        });
+      } else {
+        _client.command(['watch', root], (error, resp) => {
+          if (error) {
+            reject(error);
+          } else {
+            watcherInfo.root = root;
+            watcherInfo.relativePath = '';
+            resolve(resp);
+          }
+        });
+      }
+    });
+    
+    return promise;
+  }
+  
+  /**
+   * Issue the 'clock' command to get the time value for use with the 'since'
+   * option during 'subscribe'.
+   */
+  function clock(subscription) {
+    var self = this;
+
+    let promise = new Promise((resolve, reject) => {
+      let watcherInfo = getWatcherInfo(subscription);
+
+      _client.command(['clock', watcherInfo.root], (error, resp) => {
+        if (error) {
+          reject(error);
+        } else {
+          watcherInfo.since = resp.clock;
+          resolve(resp);
+        }
+      });
+    });
+
+    return promise;
+  }
+
+  /**
+   * Called from WatchmanWatcher (or this object during reconnect) to create
+   * a watcherInfo entry in our _watcherMap and issue a 'subscribe' to the
+   * watchman.Client, to be handled in this object.
+   */
+  function subscribe(watchmanWatcher, root) {
```

Comment:
as implemented, I believe there is a `race`, specifically if `subscribe` is invoked before `_clientPromise` settles, I believe  the `watch` invoked from `subcribe` will fail.

One way to fix this, is to only expose `client` internally via a promise that implicit waits for `_clientPromise`, something like: `WatchmanClient.prototype.client = function() { return this._clientPromise }`

We will also need to make sure that works correctly when:

* begin waiting for `_clientPromise`
* `watch(pathA)`
* connection dropped
* re-connect
* `_clientPromise` fulfills
* `watch(pathA)` should likely have connected to the  new client.